### ISSUE TITLE
"index out of bound exception" while reading the width and height from the input tensor 

### ISSUE THREAD

**Issue by ShubhamSureWash** [emgucv/emgutf#30]
Hi, I am trying to integrate the custom quantized Mobilenet SSD v2 model to unity. I am getting an "index out of bound exception" while reading the width and height from the input tensor using the following command.
int height = CocoSsdMobilenet._inputTensor.Dims[1];

I checked my model and it appears to be fine. Below is the screenshot of it.

![TFLite](https://user-images.githubusercontent.com/57176923/73864440-87fc0e00-4839-11ea-86e5-a9d5a1779496.png)

if I am commenting these lines and initializing height and width as 300, on clicking on the play button unity engine is crashing with giving any exception. Please suggest.

Our quantized TFLite model works well in the sample project provided in Android Studio by TensorFlow but its giving exception here.

**Comment by emgucv** [comment]
Can you provide the model and test code to reproduce the issue?

As per following ticket:
https://github.com/tensorflow/tensorflow/issues/22532
tflite on Windows do not implement quantization and is not able to load quantized model. I believe Quantized model is only supported for macOS, Linux, Android and iOS. If you are using TF Lite with Unity on Windows editor you may not be able to load a quantized model. Try to develop on with Unity on MacOS instead.

**Comment by ShubhamSureWash** [comment]
Hi team,

Below are the attached sample program used for running tflite model. When I
am trying with float model its working fine. I tried the quantised model in
mac but then also the same error message. detectq.tflite is my quantised
model.

Thanks,
Shubham Pandey
Vision Engineer
Surewash Ltd.
 FLoat_Model_Program_Working.zip
<https://drive.google.com/a/surewash.com/file/d/1cntYYjxb4JD1P2ItW48gfc_UrpZK09aY/view?usp=drive_web>
 Quantised_Model_Not_Working.zip
<https://drive.google.com/a/surewash.com/file/d/15qD9li7Ed6Us6PFt_lmdJ6kq47WdRZrI/view?usp=drive_web>

**Comment by ctheodorak** [comment]
Hi,
I have just trained a float model ssd_mobilenet_v1 on a custom datatset
but when i integrated it on your cocossdmobilent scene (in your script), i get false results why is that happenig

https://github.com/ctheodorak/DemoTflite/tree/master/samples
he is my model,

**Comment by emgucv** [comment]
Can you be more specific?

This is similar to asking: I used you library and other components to build a product, but my product doesn't work. Why is it happening? ... we would not be able to provide a meaning answer.

If you find a bug in a function, please point out the specific name of the function, provide the given input, what the function is returning right now and the expected output. We will be able to take a look.

If you have a working tflite code in python, along with the model and demo images, we can provide consulting service to convert python code to C# for a fee.

**Comment by ctheodorak** [comment]
Ok here is my model and labelmap https://github.com/ctheodorak/DemoTflite/tree/master/samples/New%20folder
it is a mobilenet_v1_float model training set is some statues and some coco dataset images i labeled them as other

I want it to detect statues
https://drive.google.com/open?id=1XSRLv4aeeEULnoFaoOWkapyZXBbYqCWk
here is an apk (from android studio that make it work)
Below some test images
![IMG_20191105_144921](https://user-images.githubusercontent.com/59735267/74106519-5b653080-4b70-11ea-8146-99de702bcecb.jpg)

![2](https://user-images.githubusercontent.com/59735267/74106514-50120500-4b70-11ea-99d9-285d02b5575c.jpg)


I used you Cocossdmobilenet scene with the same code


my model
![image](https://user-images.githubusercontent.com/59735267/74106856-f3641980-4b72-11ea-80b7-cc61a62654a4.png)

![image](https://user-images.githubusercontent.com/59735267/74106862-fc54eb00-4b72-11ea-9ce6-7375837dad63.png)
if is use this code
![image](https://user-images.githubusercontent.com/59735267/74106870-1e4e6d80-4b73-11ea-880c-b3710dcf17c1.png)
it should work right
the dimesnion 600X00 is parsed
i get this, no statue
![image](https://user-images.githubusercontent.com/59735267/74106896-62da0900-4b73-11ea-8b91-ab30851ac194.png)

**Comment by ctheodorak** [comment]
![image](https://user-images.githubusercontent.com/59735267/74106911-8d2bc680-4b73-11ea-8bd1-e4d5e282d4a5.png)
the code

**Comment by ctheodorak** [comment]
The model i gave you https://github.com/ctheodorak/DemoTflite/tree/master/samples/New%20folder
it works fine (it is a float model) with this code on pc
![image](https://user-images.githubusercontent.com/59735267/74150769-4ee5e400-4c13-11ea-82f9-923d49b05336.png)
I get nice results, but if i build an apk, on android it detects nothing
Is it a problem that is a float model and not a quantized one??

**Comment by emgucv** [comment]
@ShubhamSureWash 
Replying to your original post. I have downloaded  the following model used in Google's Android demo:
http://storage.googleapis.com/download.tensorflow.org/models/tflite/coco_ssd_mobilenet_v1_1.0_quant_2018_06_29.zip
Read the dimension of the input and output tensors using the following code:

```
        private static String IntArrayToString(int[] values)
        {
            return String.Format("[{0}]", String.Join(",", values));
        }

        private static String GetModelInfo(String fileName)
        {
            try
            {
                StringBuilder modelResult = new StringBuilder();
                modelResult.Append(String.Format("File Name:{0}{1}", fileName, Environment.NewLine));
                //string contents = System.Text.Encoding.UTF8.GetString(fileData.DataArray);
                using (FlatBufferModel fbm = new FlatBufferModel(fileName))
                {
                    if (!fbm.CheckModelIdentifier())
                        throw new Exception("Model identifier check failed");

                    using (Interpreter interpreter = new Interpreter(fbm))
                    {
                        Status allocateTensorStatus = interpreter.AllocateTensors();
                        if (allocateTensorStatus == Status.Error)
                            throw new Exception("Failed to allocate tensor");
                        int[] input = interpreter.InputIndices;
                        for (int i = 0; i < input.Length; i++)
                        {
                            Tensor inputTensor = interpreter.GetTensor(input[i]);

                            modelResult.Append(String.Format("Input {0} ({1}): type-{2}; dim-{3}{4}", i, inputTensor.Name,
                                inputTensor.Type, IntArrayToString( inputTensor.Dims ), Environment.NewLine));
                        }

                        int[] output = interpreter.OutputIndices;
                        for (int i = 0; i < output.Length; i++)
                        {
                            Tensor outputTensor = interpreter.GetTensor(output[i]);

                            modelResult.Append(String.Format("Output {0} ({1}): type-{2}; dim-{3}{4}", i, outputTensor.Name,
                                outputTensor.Type, IntArrayToString( outputTensor.Dims ), Environment.NewLine));
                        }

                    }
                }

                return modelResult.ToString();
            }
            catch (Exception ex)
            {
                return String.Format("Exception processing file {0}: {1} ", fileName, ex.ToString());

            }
        }
```
and got the output:
```
File Name:detect.tflite
Input 0 (normalized_input_image_tensor): type-UInt8; dim-[1,300,300,3]
Output 0 (TFLite_Detection_PostProcess): type-Float32; dim-[1,10,4]
Output 1 (TFLite_Detection_PostProcess:1): type-Float32; dim-[1,10]
Output 2 (TFLite_Detection_PostProcess:2): type-Float32; dim-[1,10]
Output 3 (TFLite_Detection_PostProcess:3): type-Float32; dim-[1]
```
I don't have the "index out of bound exception" reported in your original post.

**Comment by ShubhamSureWash** [comment]
It works fine with model provided with google sample but when I use it with my custom model it gives exception stated above. My model works fine in Sample project provided by TensorFlow in Android and iOS but it fails when tried to integrate with EMGU. In the above comment I have attached the sample code which I am using and also the my model. Emgu is working fine with my float model but its failing with my quantised model. 

Dose EMGU Unity plugin 1.13 supports TensorFlow 1.14 ? I have trained my model in TensorFlow 1.14.

**Comment by ctheodorak** [comment]
@ShubhamSureWash
I had the same problem 
![image](https://user-images.githubusercontent.com/59735267/74228224-b873fa00-4cc8-11ea-867b-7a57e743cbb0.png)
I took only an array with one element when integrate mobilenet_v2_quantized_coco instead of [1,height,widht,3]
You wont have the same problem with mobilenet_v1_quantized, i trained one and i get the results i want

**Comment by ctheodorak** [comment]
@emgucv

How i can make my scene run faster?
When i open devices's came the frames are very low cause of model's result in update function
How can i increase the fps?

**Comment by emgucv** [comment]
@ShubhamSureWash Can you use the code I pasted above on your custom model? Paste the output here so we can see. I suspect your model's input tensor's only have a single dimension. Accessing the second or 3rd dimension will results in exception.

**Comment by emgucv** [comment]
@ctheodorak Profile your code, check how much time it is spent on preparing data for tflite and how much time is spend on inference. If inference cost the most time. Look up on tensorflow or tflite on how to speed up your model.

**Comment by ctheodorak** [comment]
I train the mobilenet_v2 quantized and i used your code
![image](https://user-images.githubusercontent.com/59735267/74412620-84bede80-4e46-11ea-9885-53ee59ddceda.png)
I took this as an output
"File Name:C:/Users/user/AppData/LocalLow/DefaultCompany/Detector\\detect.tflite\r\nInput 0 (BoxPredictor_0/ClassPredictor/Conv2D_bias): type-Int32; dim-[15]\r\nOutput 0 (TFLite_Detection_PostProcess): type-Float32; dim-[1,10,4]\r\nOutput 1 (TFLite_Detection_PostProcess:1): type-Float32; dim-[1,10]\r\nOutput 2 (TFLite_Detection_PostProcess:2): type-Float32; dim-[1,10]\r\nOutput 3 (TFLite_Detection_PostProcess:3): type-Float32; dim-[1]\r\n"

and not
File Name:detect.tflite
Input 0 (normalized_input_image_tensor): type-UInt8; dim-[1,600,600,3]
Output 0 (TFLite_Detection_PostProcess): type-Float32; dim-[1,10,4]
Output 1 (TFLite_Detection_PostProcess:1): type-Float32; dim-[1,10]
Output 2 (TFLite_Detection_PostProcess:2): type-Float32; dim-[1,10]
Output 3 (TFLite_Detection_PostProcess:3): type-Float32; dim-[1]
which was supposed
BUT look at this python script
`######## Webcam Object Detection Using Tensorflow-trained Classifier #########
#
# Author: Evan Juras
# Date: 9/28/19
# Description: 
# This program uses a TensorFlow Lite object detection model to perform object 
# detection on an image or a folder full of images. It draws boxes and scores 
# around the objects of interest in each image.
#
# This code is based off the TensorFlow Lite image classification example at:
# https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/python/label_image.py
#
# I added my own method of drawing boxes and labels using OpenCV.

# Import packages
import os
import argparse
import cv2
import numpy as np
import sys
import glob
import importlib.util

# Define and parse input arguments
parser = argparse.ArgumentParser()
parser.add_argument('--modeldir', help='Folder the .tflite file is located in',
                    required=True)
parser.add_argument('--graph', help='Name of the .tflite file, if different than detect.tflite',
                    default='detect.tflite')
parser.add_argument('--labels', help='Name of the labelmap file, if different than labelmap.txt',
                    default='labelmap.txt')
parser.add_argument('--threshold', help='Minimum confidence threshold for displaying detected objects',
                    default=0.5)
parser.add_argument('--image', help='Name of the single image to perform detection on. To run detection on multiple images, use --imagedir',
                    default=None)
parser.add_argument('--imagedir', help='Name of the folder containing images to perform detection on. Folder must contain only images.',
                    default=None)
parser.add_argument('--edgetpu', help='Use Coral Edge TPU Accelerator to speed up detection',
                    action='store_true')

args = parser.parse_args()

MODEL_NAME = args.modeldir
GRAPH_NAME = args.graph
LABELMAP_NAME = args.labels
min_conf_threshold = float(args.threshold)
use_TPU = args.edgetpu

# Parse input image name and directory. 
IM_NAME = args.image
IM_DIR = args.imagedir

# If both an image AND a folder are specified, throw an error
if (IM_NAME and IM_DIR):
    print('Error! Please only use the --image argument or the --imagedir argument, not both. Issue "python TFLite_detection_image.py -h" for help.')
    sys.exit()

# If neither an image or a folder are specified, default to using 'test1.jpg' for image name
if (not IM_NAME and not IM_DIR):
    IM_NAME = 'test1.jpg'

# Import TensorFlow libraries
# If tensorflow is not installed, import interpreter from tflite_runtime, else import from regular tensorflow
# If using Coral Edge TPU, import the load_delegate library
pkg = importlib.util.find_spec('tensorflow')
if pkg is None:
    from tflite_runtime.interpreter import Interpreter
    if use_TPU:
        from tflite_runtime.interpreter import load_delegate
else:
    from tensorflow.lite.python.interpreter import Interpreter
    if use_TPU:
        from tensorflow.lite.python.interpreter import load_delegate

# If using Edge TPU, assign filename for Edge TPU model
if use_TPU:
    # If user has specified the name of the .tflite file, use that name, otherwise use default 'edgetpu.tflite'
    if (GRAPH_NAME == 'detect.tflite'):
        GRAPH_NAME = 'edgetpu.tflite'


# Get path to current working directory
CWD_PATH = os.getcwd()

# Define path to images and grab all image filenames
if IM_DIR:
    PATH_TO_IMAGES = os.path.join(CWD_PATH,IM_DIR)
    images = glob.glob(PATH_TO_IMAGES + '/*')

elif IM_NAME:
    PATH_TO_IMAGES = os.path.join(CWD_PATH,IM_NAME)
    images = glob.glob(PATH_TO_IMAGES)

# Path to .tflite file, which contains the model that is used for object detection
PATH_TO_CKPT = os.path.join(CWD_PATH,MODEL_NAME,GRAPH_NAME)

# Path to label map file
PATH_TO_LABELS = os.path.join(CWD_PATH,MODEL_NAME,LABELMAP_NAME)

# Load the label map
with open(PATH_TO_LABELS, 'r') as f:
    labels = [line.strip() for line in f.readlines()]

# Have to do a weird fix for label map if using the COCO "starter model" from
# https://www.tensorflow.org/lite/models/object_detection/overview
# First label is '???', which has to be removed.
if labels[0] == '???':
    del(labels[0])

# Load the Tensorflow Lite model.
# If using Edge TPU, use special load_delegate argument
if use_TPU:
    interpreter = Interpreter(model_path=PATH_TO_CKPT,
                              experimental_delegates=[load_delegate('libedgetpu.so.1.0')])
    print(PATH_TO_CKPT)
else:
    interpreter = Interpreter(model_path=PATH_TO_CKPT)

interpreter.allocate_tensors()

# Get model details
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
height = input_details[0]['shape'][1]
width = input_details[0]['shape'][2]

floating_model = (input_details[0]['dtype'] == np.float32)

input_mean = 127.5
input_std = 127.5

# Loop over every image and perform detection
for image_path in images:

    # Load image and resize to expected shape [1xHxWx3]
    image = cv2.imread(image_path)
    image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    imH, imW, _ = image.shape 
    image_resized = cv2.resize(image_rgb, (width, height))
    input_data = np.expand_dims(image_resized, axis=0)

    # Normalize pixel values if using a floating model (i.e. if model is non-quantized)
    if floating_model:
        input_data = (np.float32(input_data) - input_mean) / input_std
    print(input_details)

    # Perform the actual detection by running the model with the image as input
    interpreter.set_tensor(input_details[0]['index'],input_data)
    interpreter.invoke()

    # Retrieve detection results
    boxes = interpreter.get_tensor(output_details[0]['index'])[0] # Bounding box coordinates of detected objects
    classes = interpreter.get_tensor(output_details[1]['index'])[0] # Class index of detected objects
    scores = interpreter.get_tensor(output_details[2]['index'])[0] # Confidence of detected objects
    #num = interpreter.get_tensor(output_details[3]['index'])[0]  # Total number of detected objects (inaccurate and not needed)

    # Loop over all detections and draw detection box if confidence is above minimum threshold
    for i in range(len(scores)):
        if ((scores[i] > min_conf_threshold) and (scores[i] <= 1.0)):

            # Get bounding box coordinates and draw box
            # Interpreter can return coordinates that are outside of image dimensions, need to force them to be within image using max() and min()
            ymin = int(max(1,(boxes[i][0] * imH)))
            xmin = int(max(1,(boxes[i][1] * imW)))
            ymax = int(min(imH,(boxes[i][2] * imH)))
            xmax = int(min(imW,(boxes[i][3] * imW)))
            
            cv2.rectangle(image, (xmin,ymin), (xmax,ymax), (10, 255, 0), 2)

            # Draw label
            object_name = labels[int(classes[i])] # Look up object name from "labels" array using class index
            label = '%s: %d%%' % (object_name, int(scores[i]*100)) # Example: 'person: 72%'
            labelSize, baseLine = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, 0.7, 2) # Get font size
            label_ymin = max(ymin, labelSize[1] + 10) # Make sure not to draw label too close to top of window
            cv2.rectangle(image, (xmin, label_ymin-labelSize[1]-10), (xmin+labelSize[0], label_ymin+baseLine-10), (255, 255, 255), cv2.FILLED) # Draw white box to put label text in
            cv2.putText(image, label, (xmin, label_ymin-7), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 0, 0), 2) # Draw label text

    # All the results have been drawn on the image, now display the image
    cv2.imshow('Object detector', image)

    # Press any key to continue to next image, or press 'q' to quit
    if cv2.waitKey(0) == ord('q'):
        break

# Clean up
cv2.destroyAllWindows()
`
It gives this output
![image](https://user-images.githubusercontent.com/59735267/74412805-e2ebc180-4e46-11ea-9682-7afde2768328.png)
My model :https://github.com/ctheodorak/DemoTflite/tree/master/MobilenetV2_qunatized
And also it works fine with the python script why i get the same exception in your plugin.
